lib/github: uses nitcorn to listen Github hook events.

Morriar opened the pull request:
This module introduces an abstract listener for Github hooks with nitcorn.

PS: OMG first PR with syntax highligh!!!! Thx @R4PaSs !

Signed-off-by: Alexandre Terrasa <alexandre@moz-code.org>

Morriar commented:
So, adding a new module, never imported by anyone break the Neo4J API according to Jenkins...

I will try a rebuild before checking the source code :p

privat commented on lib/github/hooks.nit:
```diff
@@ -0,0 +1,172 @@
+# class LogHookListener
+#	super HookListener
+#
+#	redef fun api: GithubAPI do return new GithubAPI(get_github_oauth)
```
api sert a rien ici

Morriar replied:
bien vu

privat commented on lib/github/hooks.nit:
```diff
@@ -0,0 +1,172 @@
+#
+#	redef fun api: GithubAPI do return new GithubAPI(get_github_oauth)
+#
+#	redef fun apply_event(event) do
```
peut-etre dans la doc proposer de passer par un double dispatch plutot que par un tas de if, non?

~~~
class LogHookListener
   super HookListener
   redef fun apply_event(event) do event.log_event(self)
end
redef class GithubEvent # la super-classe
   fun log_event(l: LogHookListener) is do end # by defaut do nothing
end
redef class CommitCommentEvent
   redef fun log_event(l) do des trucs...
end
~~~

Morriar replied:
Bonne idée.

privat commented on lib/github/hooks.nit:
```diff
@@ -0,0 +1,172 @@
+#
+# var api = new GithubAPI(get_github_oauth)
+# var listener = new LogHookListener(api, "127.0.0.1", 8080)
+# # listener.listen # uncomment to start listening
```
uncomment, vraiment?

Morriar replied:
Je voulais justifier la ligne en commentaire par autre chose que `# avoid infinite loop in nitunit`

privat replied:
Pas faux.  mettre un `~~~nitish` eviterait le probleme, mais il n'y aurait plus de vérification.
Ajouter un tag intermediaire entre `nitish` et `nit`. Peut-etre `nounit`?

privat replied:
Pour ajouter à la discussion, avoir un programme d'exemple en entier dans de la nitdoc est pas si pratique:

* chiant a éditer a cause du `#` et des tabulation
* ne peut se faire exécuter simplement par l'utilisateur
* peut-etre trop long dans la nitdoc.

Tout ça est sans doute lié à #852

Sans doute la bonne façon serait de mettre ces programmes d'exemple dans des fichiers à part pour que les clients puissent les exécuter ou les modifier facilement et trouver un moyen pour en faire quelque chose d'intelligent

privat commented on lib/github/hooks.nit:
```diff
@@ -0,0 +1,172 @@
+	var verbosity = 0
+
+	# Print `message` if `lvl` <= `verbosity`
+	fun message(lvl: Int, message: String) do
```
un jour un vrai logger dans la lib?

Morriar replied:
J'avoue que ça commence à être redondant.

Remarque, le besoin d'un logger dans la lib indique qu'on a de plus en plus de clients en dehors de src! À une époque le seul logger dont on avait besoin c'était ToolContext. Cool!

privat commented on lib/github/hooks.nit:
```diff
@@ -0,0 +1,172 @@
+
+	# What to do when we receive an event from a hook?
+	#
+	# This method MUST be defined in specialized listeners.
```
est-ce que le `abstract` ne dit pas déjà "This method MUST be defined in specialized subclasses"?

Morriar replied:
Avec l'absence de contraintes à la compilation sur les abstracts. J'ai pris pour acquis que rien n'est vraiment à redéfinir tant que ça n'a pas planté à l'exécution. Donc autant faire gagner une compilation en le précisant dans la doc.

privat replied:
La doc générée devrait déjà indiquer le `abstract`, non ? Forcer le programmeur à écrire dans sa doc une phrase qui veut dire exactement `abstract` n'est pas digne d'une documentation parfaite. Ou alors `abstract` n'est pas exactement synonyme de `This method MUST be defined in specialized subclasses` mais dans ce cas là on a un problème de spec et il faut définir mieux que ce que `abstract` veut dire au niveau du langage.

Le fait que le compilo soit taquin et laisse le programme se faire compiler et exécuter jusqu'à faire boom au dernier moment boom est un problème indépendant #808

privat commented on lib/github/hooks.nit:
```diff
@@ -0,0 +1,172 @@
+		else if kind == "status" then
+			return new StatusEvent.from_json(api, json)
+		else
+			return new GithubEvent.from_json(api, json)
```
je propose d'ajouter un `message("received unknown github event {kind}")` les gens du futur en seront peut-être reconnaissant.

privat commented on lib/github/hooks.nit:
```diff
@@ -0,0 +1,172 @@
+		if not request.header.has_key("X-GitHub-Event") then
+			return new HttpResponse(403)
+		end
+		var kind = request.header["X-GitHub-Event"]
```
On peut simplifier les lignes précédentes en
~~~nit
var kind = request.header.get_or_null("X-GitHub-Event")
if kind == null then return new HttpResponse(403)
~~~

privat commented:
C'est cool. et la colo facilite vraiment la relecture

Morriar commented:
Reroll:

* [x] Simplified example with double dispatch
* [x] Removed comment about abstract
* [x] Simplified access to map value with `get_or_null`

Things that will not change in this PR:

* Remove the example from the comment (we don't know where to put that for now) 
* `# uncomment` (no other way to show that information without setting it in `nitish`)
* Add a message for `unknown events` (no more needed thanks to double dispatch)
* Implement un real logger in the lib

privat commented:
+1

R4PaSs commented:
+1
